File: cli/src/commands/status.js
import path from 'path';
import fs from 'fs/promises';
import { loadGraph, loadMeta } from '../graph.js';

/**
 * Register the `status` command on the given Commander program.
 *
 * @param {import('commander').Command} program - The Commander program instance.
 */
export function registerStatusCommand(program) {
  program
    .command('status [dir]')
    .description('Show the status of the code graph for a project')
    .action(async (dir) => {
      const rootDir = path.resolve(dir || '.');
      const outputDir = path.join(rootDir, '.codemap');

      // Check if .codemap exists
      try {
        await fs.access(outputDir);
      } catch {
        console.error('No code graph found. Run "codegraph scan" first.');
        process.exit(1);
      }

      let graph, meta;
      try {
        graph = await loadGraph(outputDir);
        meta = await loadMeta(outputDir);
      } catch (err) {
        console.error(`Error loading code graph: ${err.message}`);
        process.exit(1);
      }

      // Print status
      console.log(`Project: ${graph.project.name}`);
      console.log(`Scanned at: ${graph.scannedAt}`);
      console.log(`Commit: ${graph.commitHash || '(none)'}`);
      console.log(`Files: ${graph.summary.totalFiles}`);
      console.log(`Functions: ${graph.summary.totalFunctions}`);
      console.log(`Classes: ${graph.summary.totalClasses}`);
      console.log(`Modules: ${graph.summary.modules.join(', ')}`);

      // Languages
      const langEntries = Object.entries(graph.summary.languages);
      if (langEntries.length > 0) {
        const langStr = langEntries.map(([lang, count]) => `${lang}(${count})`).join(', ');
        console.log(`Languages: ${langStr}`);
      }

      // Last update from meta
      if (meta.lastScanAt) {
        console.log(`Last update: ${meta.lastScanAt}`);
      }

      // Dirty files: files whose current hash differs from meta.fileHashes
      // (We just report what's in meta for now; actual dirty detection
      // would require re-traversing and re-hashing which is what `update` does)
      const fileHashCount = meta.fileHashes ? Object.keys(meta.fileHashes).length : 0;
      console.log(`Tracked files: ${fileHashCount}`);
    });
}

File: rust-cli/src/commands/status.rs
use clap::Args;
use std::path::PathBuf;

use crate::graph::{load_graph, load_meta};

#[derive(Args)]
pub struct StatusArgs {
    /// Project directory
    pub dir: Option<String>,
}

pub fn run(args: StatusArgs) {
    let dir = args.dir.unwrap_or_else(|| ".".to_string());
    let root_dir = match PathBuf::from(&dir).canonicalize() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("Error: cannot resolve directory '{}': {}", dir, e);
            std::process::exit(1);
        }
    };
    let output_dir = root_dir.join(".codemap");

    if !output_dir.exists() {
        eprintln!("No code graph found. Run \"codegraph scan\" first.");
        std::process::exit(1);
    }

    let graph = match load_graph(&output_dir) {
        Ok(g) => g,
        Err(e) => {
            eprintln!("Error loading code graph: {e}");
            std::process::exit(1);
        }
    };

    let meta = load_meta(&output_dir).ok();

    println!("Project: {}", graph.project.name);
    println!("Scanned at: {}", graph.scanned_at);
    println!("Commit: {}", graph.commit_hash.as_deref().unwrap_or("(none)"));
    println!("Files: {}", graph.summary.total_files);
    println!("Functions: {}", graph.summary.total_functions);
    println!("Classes: {}", graph.summary.total_classes);
    println!("Modules: {}", graph.summary.modules.join(", "));

    // 语言分布
    if !graph.summary.languages.is_empty() {
        let mut lang_entries: Vec<_> = graph.summary.languages.iter().collect();
        lang_entries.sort_by_key(|(k, _)| k.as_str());
        let lang_str: Vec<String> = lang_entries
            .iter()
            .map(|(lang, count)| format!("{lang}({count})"))
            .collect();
        println!("Languages: {}", lang_str.join(", "));
    }

    // 上次更新时间（来自 meta）
    if let Some(ref m) = meta {
        println!("Last update: {}", m.last_scan_at);
    }

    // 已追踪文件数
    let tracked = meta
        .as_ref()
        .map(|m| m.file_hashes.len())
        .unwrap_or(0);
    println!("Tracked files: {tracked}");
}
